File: samples/c/08objects/trapezoid.c
#include "trapezoid.h"

Trapezoid* createTrapezoid(Rectangle* r, Triangle* t) {
    if (r->height != t->rectangle->height) {
        return NULL;
    }
    Trapezoid* p = malloc(sizeof(Trapezoid));
    p->rectanglePart = r;
    p->trianglePart = t;
    return p;
}

float trapezoidArea(Trapezoid* p) {
    return rectangleArea(p->rectanglePart) +
        triangleArea(p->trianglePart);
}

float trapezoidPerimeter(Trapezoid* p) {
    return rectanglePerimeter(p->rectanglePart) - p->rectanglePart->height
        + trianglePerimeter(p->trianglePart) - p->trianglePart->rectangle->height;
}

float deleteTrapezoid(Trapezoid** p) {
    (*p)->rectanglePart = NULL;
    (*p)->trianglePart = NULL;
    free(*p);
    *p = NULL;
}

File: samples/java/08objects/main/Trapezoid.java
package main;

import main.Rectangle;
import main.Triangle;

public class Trapezoid extends Rectangle {
    private Triangle triangle;

    public Trapezoid(double totalWidth, double height, double triangleWidth) {
        super(totalWidth, height);
        this.triangle = new Triangle(triangleWidth, height);
    }

    public double area() {
        return super.area() - this.triangle.area();
    }

    public double perimeter() {
        return super.perimeter() - super.getHeight() - this.triangle.getWidth() + this.triangle.getHypotenuse();
    }
}